ocaml/xapi-idl/xen/device_number.ml

```diff
@@ -119,112 +106,119 @@ let rec string_of_int26 x =
   high' ^ low'
 
-module String = struct
-  include String
-
-  let fold_right f string accu =
-    let accu = ref accu in
-    for i = length string - 1 downto 0 do
-      accu := f string.[i] !accu
-    done ;
-    !accu
-
-  let explode string = fold_right (fun h t -> h :: t) string []
-
-  let implode list = concat "" (List.map (String.make 1) list)
-end
-
-(** Convert a linux device string back into an integer *)
-let int26_of_string x =
-  let ints =
-    List.map (fun c -> int_of_char c - int_of_char 'a' + 1) (String.explode x)
-  in
-  List.fold_left (fun acc x -> (acc * 26) + x) 0 ints - 1
-
-let to_linux_device =
+let to_linux_prefix = function
+  | Xen ->
+      "xvd"
+  | Scsi ->
+      "sd"
+  | Floppy ->
+      "fd"
+  | Ide ->
+      "xvd"
+
+let to_linux_device (bus, disk, part) =
   let p x = if x = 0 then "" else string_of_int x in
-  function
-  | Xen, disk, part ->
-      Printf.sprintf "xvd%s%s" (string_of_int26 disk) (p part)
-  | Scsi, disk, part ->
-      Printf.sprintf "sd%s%s" (string_of_int26 disk) (p part)
-  | Floppy, disk, part ->
-      Printf.sprintf "fd%s%s" (string_of_int26 disk) (p part)
-  | Ide, disk, part ->
-      Printf.sprintf "xvd%s%s" (string_of_int26 disk) (p part)
+  let bus = to_linux_prefix bus in
+  Printf.sprintf "%s%s%s" bus (string_of_int26 disk) (p part)
 
 let of_linux_device x =
-  let letter c = 'a' <= c && c <= 'z' in
-  let digit c = '0' <= c && c <= '9' in
-  let take f x =
-    let rec inner f acc = function
-      | x :: xs ->
-          if f x then inner f (x :: acc) xs else (List.rev acc, x :: xs)
-      | [] ->
-          (List.rev acc, [])
+  let open Astring in
+  let b26_to_int x =
+    (* Convert a linux device string back into an integer *)
+    (* Assumes all characters are in range *)
+    let b26 =
+      String.Sub.to_string x
+      |> Stdlib.String.to_seq
+      |> Seq.map (fun c -> int_of_char c - int_of_char 'a' + 1)
+      |> Seq.fold_left (fun acc x -> (acc * 26) + x) 0
     in
-    inner f [] x
+    b26 - 1
+  in
+
+  let parse_int x =
+    match String.Sub.span ~min:1 ~sat:Char.Ascii.is_digit x with
+    | i, s ->
+        Option.map (fun i -> (i, s)) (String.Sub.to_int i)
+  in
+  let parse_b26 x =
+    match String.Sub.span ~min:1 ~sat:Char.Ascii.is_lower x with
+    | b, s ->
+        (b26_to_int b, s)
   in
   (* Parse a string "abc123" into x, y where x is "abc" interpreted as base-26
      and y is 123 *)
   let parse_b26_int x =
-    let d, p = take letter x in
-    let d' = int26_of_string (String.implode d) in
-    let p' = if p = [] then 0 else int_of_string (String.implode p) in
-    (d', p')
+    let pre, x = parse_b26 x in
+    if String.Sub.is_empty x then
+      Some (pre, 0)
+    else
+      let* post, x = parse_int x in
+      if not (String.Sub.is_empty x) then
+        None
+      else
+        Some (pre, post)
   in
   (* Parse a string "123p456" into x, y where x = 123 and y = 456 *)
```

Comment:
Why not use `sscanf` with `%dp%d`? A lot of this parsing code looks too complicated.